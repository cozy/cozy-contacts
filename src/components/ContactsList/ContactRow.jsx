import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'

import { models } from 'cozy-client'
import withBreakpoints from 'cozy-ui/transpiled/react/helpers/withBreakpoints'
import { TableRow } from 'cozy-ui/transpiled/react/Table'

import { fullContactPropTypes } from '../ContactPropTypes'
import withModal from '../HOCs/withModal'
import ContactCardModal from '../Modals/ContactCardModal'
import ContactWithSelection from './ContactSelection'
import ContactPhone from './Contacts/ContactPhone'
import ContactIdentity from './Contacts/ContactIdentity'
import ContactCozy from './Contacts/ContactCozy'
import ContactEmail from './Contacts/ContactEmail'

const { getPrimaryCozy, getPrimaryPhone, getPrimaryEmail } = models.contact

/**
 * Returns whether a document has been updated
 * @param {object} document - Initial document
 * @param {object} nextDocument - Potentially modified document
 * @returns {boolean}
 */
export const hasDocBeenUpdated = (document, nextDocument) =>
  document._rev !== nextDocument._rev ||
  get(document, 'cozyMetadata.updatedAt') !==
    get(nextDocument, 'cozyMetadata.updatedAt')

class ContactRow extends Component {
  shouldComponentUpdate(nextProps) {
    const { contact, ...otherProps } = this.props

    const hasOtherPropsBeenChanged = Object.entries(otherProps).some(
      ([key, otherProp]) => {
        return otherProp !== nextProps[key]
      }
    )

    if (
      hasDocBeenUpdated(contact, nextProps.contact) ||
      hasOtherPropsBeenChanged
    ) {
      return true
    }
    return false
  }

  render() {
    const {
      contact,
      showModal,
      breakpoints: { isMobile }
    } = this.props
    const email = getPrimaryEmail(contact) || undefined
    const phone = getPrimaryPhone(contact) || undefined
    const cozyUrl = getPrimaryCozy(contact) || undefined

    return (
      <TableRow
        className="contact"
        data-testid="contact-row"
        onClick={() => showModal(<ContactCardModal id={contact._id} />)}
      >
        <ContactWithSelection contact={contact} />
        <ContactIdentity contact={contact} />
        {!isMobile && <ContactEmail email={email} />}
        {!isMobile && <ContactPhone phone={phone} />}
        {!isMobile && <ContactCozy cozyUrl={cozyUrl} />}
      </TableRow>
    )
  }
}

ContactRow.propTypes = {
  contact: fullContactPropTypes.isRequired,
  showModal: PropTypes.func.isRequired
}
ContactRow.defaultProps = {
  selection: null
}

export default withBreakpoints()(withModal(ContactRow))
