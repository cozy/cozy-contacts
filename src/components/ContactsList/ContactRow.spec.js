import React from 'react'
import { mount } from 'enzyme'

import ContactRow from './ContactRow'

describe('ContactRow', () => {
  test('should accept the strict minimum', () => {
    const contact = { email: [{ address: 'johndoe@localhost' }] }
    const contactRowInstance = <ContactRow contact={contact} />
    const contactrow = mount(contactRowInstance)
    const contactrowemail = contactrow.find('ContactEmail')
    expect(contactrowemail).toBeDefined()
    expect(contactrowemail.text()).toBe(contact.email[0].address)
  })
  test('should display data', () => {
    const contact = {
      name: { familyName: 'Doe', givenName: 'John' },
      phone: [{ number: '0123456789' }],
      email: [{ address: 'johndoe@localhost' }]
    }
    const contactRowInstance = <ContactRow contact={contact} />
    const contactrow = mount(contactRowInstance)
    const contactrowname = contactrow.find('ContactName')
    expect(contactrowname).toBeDefined()
    expect(contactrowname.text()).toEqual(
      expect.stringContaining(contact.name.givenName)
    )
    expect(contactrowname.text()).toEqual(
      expect.stringContaining(contact.name.familyName)
    )
    const contactrowphone = contactrow.find('ContactPhone')
    expect(contactrowphone).toBeDefined()
    expect(contactrowphone.text()).toBe(contact.phone[0].number)
  })
  test('should display empty string for missing information', () => {
    const contact = { email: [{ address: 'johndoe@localhost' }] }
    const contactRowInstance = <ContactRow contact={contact} />
    const contactrow = mount(contactRowInstance)
    const contactrowname = contactrow.find('ContactName')
    expect(contactrowname).toBeDefined()
    expect(contactrowname.text().trim()).toBe('')
    const contactrowphone = contactrow.find('ContactPhone')
    expect(contactrowphone).toBeDefined()
    expect(contactrowphone.text().trim()).toBe('—')
  })
  test('should accept empty array', () => {
    const contact = { email: [] }
    const contactRowInstance = <ContactRow contact={contact} />
    const contactrow = mount(contactRowInstance)
    const contactrowemail = contactrow.find('ContactEmail')
    expect(contactrowemail).toBeDefined()
    expect(contactrowemail.text()).toBe('—')
  })
})
