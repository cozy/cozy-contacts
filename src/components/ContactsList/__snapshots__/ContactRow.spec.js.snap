// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactRow should match the contact snapshot 1`] = `
<div
  className="styles__TableRow___1D62H contact"
  data-testid="contact-row"
  onClick={[Function]}
>
  <div
    className="contact-selection"
    onClick={[Function]}
  >
    <span
      data-input="checkbox"
    >
      <input
        checked={false}
        readOnly={true}
        type="checkbox"
      />
      <label />
    </span>
  </div>
  <div
    className="contact-identity"
  >
    <div
      className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
      style={
        Object {
          "--circleSize": "32px",
          "backgroundColor": "var(--brightSun)",
        }
      }
    >
      
      <span
        className="styles__c-avatar-initials___310qC"
      >
        JD
      </span>
    </div>
    <div
      className="u-ellipsis u-ml-1"
    >
      <span
        className=""
      >
        John
         
      </span>
      <span
        className="u-fw-bold"
      >
        Doe
         
      </span>
    </div>
  </div>
  <div
    className="contact-email u-ellipsis"
  >
    johndoe@localhost
  </div>
  <div
    className="contact-phone u-ellipsis"
  >
    0123456789
  </div>
  <div
    className="contact-cozyurl u-ellipsis"
  >
    http://johndoe.mycozy.cloud
  </div>
</div>
`;
